# 01-Intro Homework Submission

## Q1. Pandas version
What's the version of Pandas that you installed?

You can get the version information using the ```__version__``` field:

In [27]:
import pandas as pd
import numpy as np

pd.__version__

'2.3.2'

The pandas version I installed is **2.3.2**.

## Getting the data
For this homework, we'll use the Car Fuel Efficiency dataset.

wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

In [3]:
df = pd.read_csv("https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv")
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

## Q2. Records count
How many records are in the dataset?

- 4704
- 8704
- 9704
- 17704

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


The dataset has **9704** records.

## Q3. Fuel types
How many fuel types are presented in the dataset?

- 1
- 2
- 3
- 4

In [7]:
#Check for the unique fuel_type in the dataset.
df["fuel_type"].nunique()

2

There are 2 unique fuel types in the dataset.

## Q4. Missing values
How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3
- 4

In [9]:
#Chech for null values in each column in the dataset
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

The dataset has 4 columns with missing values; **num_cylinders, horsepower, acceleration, and num_doors.**

## Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

- 13.75
- 23.75
- 33.75
- 43.75

In [16]:
#Subset the data to origin 'Asia' only
df_Asia = df[df["origin"] == 'Asia']

#Find max fuel efficiency for asian cars
max_fuel_efficiency_asia = df_Asia["fuel_efficiency_mpg"].max()

#Print the results
print(round(max_fuel_efficiency_asia, 2))

23.76


Asian cars have a maximum fuel efficieny of **23.76**.

## Q6. Median value of horsepower
1. Find the median value of horsepower column in the dataset.
2. Next, calculate the most frequent value of the same horsepower column.
3. Use fillna method to fill the missing values in horsepower column with the most frequent value from the previous step.
4. Now, calculate the median value of horsepower once again.

Has it changed?

- Yes, it increased
- Yes, it decreased
- No

In [17]:
#Median df["horsepower"]
median_horsepower = df["horsepower"].median()
print(median_horsepower)

149.0


The median horsepower is **149.0**.

In [20]:
#Mode of df["horsepower"]
most_frequent_horsepower = df["horsepower"].mode()[0]
print(most_frequent_horsepower)

152.0


The most frequent horsepower is **152.0**.

In [21]:
#Fill null values in df["horsepower"] with mode
df["horsepower"].fillna(most_frequent_horsepower)

#Median df["horsepower"] after filling nulls with mode
new_median_horsepower = df["horsepower"].median()

print(new_median_horsepower)

149.0


The new median horsepower without null values is **149.0**.

No. The value has not changed. It is still **149.0**

## Q7. Sum of weights
1. Select all the cars from Asia
2. Select only columns vehicle_weight and model_year
3. Select the first 7 values
4. Get the underlying NumPy array. Let's call it X.
5. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
6. Invert XTX.
7. Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
8. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
9. What's the sum of all the elements of the result?

In [23]:
#Slelect all cars from Asia
df_asia = df[df["origin"] == 'Asia']
df_asia.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                   250            1.0       174.0     2714.219310   
12                  320            5.0       145.0     2783.868974   
14                  200            6.0       160.0     3582.687368   
20                  150            3.0       197.0     2231.808142   
21                  160            4.0       133.0     2659.431451   

    acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8           10.3        2016   Asia    Diesel  Front-wheel drive       -1.0   
12          15.1        2010   Asia    Diesel    All-wheel drive        1.0   
14          14.9        2007   Asia    Diesel    All-wheel drive        0.0   
20          18.7        2011   Asia  Gasoline  Front-wheel drive        1.0   
21           NaN        2016   Asia  Gasoline  Front-wheel drive       -1.0   

    fuel_efficiency_mpg  
8             16.823554  
12            16.175820  
14            11.871091  
20            18.889083  
21            16.077730

In [26]:
#Select only columns; vehicle_weight and model_year 
df_subset_asia = df_asia[["vehicle_weight", "model_year"]]

#Slect the first seven values 
df_subset_asia_head  = df_subset_asia.head(7)

# Convert to NumPy array
X = df_subset_asia_head.to_numpy()
print(X)

[[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]


In [28]:
#Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
XTX = np.dot(X.T, X)

#Invert XTX.
XTX_inv = np.linalg.inv(XTX)

print("XTX:\n", XTX)
print("Inverse of XTX:\n", XTX_inv)

XTX:
 [[62248334.33150762 41431216.5073268 ]
 [41431216.5073268  28373339.        ]]
Inverse of XTX:
 [[ 5.71497081e-07 -8.34509443e-07]
 [-8.34509443e-07  1.25380877e-06]]


In [29]:
#Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

#Multiply the inverse of XTX with the transpose of X,
XTX_inv_XT = np.dot(XTX_inv, X.T)

#Then multiply the result by y. Call the result w.
w = np.dot(XTX_inv_XT, y)

print(w)

[0.01386421 0.5049067 ]


In [32]:
#The sum of all the elements of the result, w
total = np.sum(w)
print("Sum of all elements in w:", total)

Sum of all elements in w: 0.5187709081074016


The sum of all the elements of the result is **0.5187709081074016**.